src/server/editorServices.ts

```diff
@@ -541,174 +576,162 @@ namespace ts.server {
-            if (shouldRefreshInferredProjects) {
-                this.refreshInferredProjects();
-            }
         }
 
-        private onSourceFileChanged(fileName: NormalizedPath) {
+        private onSourceFileChanged(fileName: NormalizedPath, eventKind: FileWatcherEventKind) {
             const info = this.getScriptInfoForNormalizedPath(fileName);
             if (!info) {
                 this.logger.info(`Error: got watch notification for unknown file: ${fileName}`);
                 return;
             }
 
-            if (!this.host.fileExists(fileName)) {
+            if (eventKind === FileWatcherEventKind.Deleted) {
                 // File was deleted
                 this.handleDeletedFile(info);
             }
             else {
-                if (info && (!info.isScriptOpen())) {
+                if (!info.isScriptOpen()) {
                     if (info.containingProjects.length === 0) {
-                        // Orphan script info, remove it as we can always reload it on next open
-                        info.stopWatcher();
+                        // Orphan script info, remove it as we can always reload it on next open file request
+                        this.stopWatchingScriptInfo(info, WatcherCloseReason.OrphanScriptInfoWithChange);
                         this.filenameToScriptInfo.delete(info.path);
                     }
                     else {
                         // file has been changed which might affect the set of referenced files in projects that include
                         // this file and set of inferred projects
                         info.reloadFromFile();
-                        this.updateProjectGraphs(info.containingProjects);
+                        this.delayUpdateProjectGraphs(info.containingProjects);
                     }
                 }
             }
         }
 
         private handleDeletedFile(info: ScriptInfo) {
-            this.logger.info(`${info.fileName} deleted`);
-
-            info.stopWatcher();
+            this.stopWatchingScriptInfo(info, WatcherCloseReason.FileDeleted);
 
             // TODO: handle isOpen = true case
 
             if (!info.isScriptOpen()) {
                 this.filenameToScriptInfo.delete(info.path);
-                this.lastDeletedFile = info;
 
                 // capture list of projects since detachAllProjects will wipe out original list
                 const containingProjects = info.containingProjects.slice();
 
                 info.detachAllProjects();
 
                 // update projects to make sure that set of referenced files is correct
-                this.updateProjectGraphs(containingProjects);
-                this.lastDeletedFile = undefined;
+                this.delayUpdateProjectGraphs(containingProjects);
 
-                if (!this.eventHandler) {
-                    return;
-                }
+                // TODO: (sheetalkamat) Someway to send this event so that error checks are updated?

```

Comment:
Its todo because I need to fix this before merging. I will be updating this soon